File: client/src/main/java/su/plo/voice/client/audio/device/BaseAudioDevice.java
package su.plo.voice.client.audio.device;

import com.google.common.collect.ListMultimap;
import com.google.common.collect.Multimaps;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.Nullable;
import su.plo.voice.api.client.PlasmoVoiceClient;
import su.plo.voice.api.client.audio.device.AudioDevice;
import su.plo.voice.api.client.audio.device.DeviceException;
import su.plo.voice.api.client.audio.filter.AudioFilter;
import su.plo.voice.api.util.Params;

import javax.sound.sampled.AudioFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.function.Predicate;

@RequiredArgsConstructor
public abstract class BaseAudioDevice implements AudioDevice {

    protected final PlasmoVoiceClient client;
    protected final @Nullable String name;

    protected AudioFormat format;
    protected Params params;
    @Getter
    protected int bufferSize;

    protected final ListMultimap<AudioFilter.Priority, AudioFilter> filters = Multimaps.synchronizedListMultimap(
            Multimaps.newListMultimap(new HashMap<>(), ArrayList::new)
    );

    @Override
    public void reload() throws DeviceException {
        if (!isOpen()) throw new DeviceException("Device is not open");

        close();
        open(format, params);
    }

    @Override
    public void addFilter(AudioFilter filter, AudioFilter.Priority priority) {
        // check if filter already exists
        for (AudioFilter.Priority filtersPriority : AudioFilter.Priority.values()) {
            Collection<AudioFilter> filters = this.filters.get(filtersPriority);
            if (filters.contains(filter)) {
                throw new IllegalArgumentException("Filter is already exist with priority: " + filtersPriority);
            }
        }

        filters.put(priority, filter);
    }

    @Override
    public void addFilter(AudioFilter filter) {
        addFilter(filter, AudioFilter.Priority.NORMAL);
    }

    @Override
    public void removeFilter(AudioFilter filter) {
        for (AudioFilter.Priority filtersPriority : AudioFilter.Priority.values()) {
            Collection<AudioFilter> filters = this.filters.get(filtersPriority);
            filters.remove(filter);
        }
    }

    @Override
    public Collection<AudioFilter> getFilters() {
        return filters.values();
    }

    @Override
    public short[] processFilters(short[] samples, @Nullable Predicate<AudioFilter> excludeFilter) {
        for (AudioFilter filter : filters.values()) {
            if (filter.isEnabled() && (excludeFilter == null || !excludeFilter.test(filter))) {
                samples = filter.process(samples);
            }
        }

        return samples;
    }
}

File: client/src/main/kotlin/su/plo/voice/client/audio/device/BaseAudioDevice.kt
package su.plo.voice.client.audio.device

import com.google.common.collect.ListMultimap
import com.google.common.collect.Multimaps
import lombok.RequiredArgsConstructor
import su.plo.voice.api.client.PlasmoVoiceClient
import su.plo.voice.api.client.audio.device.AudioDevice
import su.plo.voice.api.client.audio.device.DeviceException
import su.plo.voice.api.client.audio.filter.AudioFilter
import su.plo.voice.api.client.audio.filter.AudioFilterContext
import java.util.function.Predicate
import javax.sound.sampled.AudioFormat

@RequiredArgsConstructor
abstract class BaseAudioDevice(
    protected val voiceClient: PlasmoVoiceClient,
    override val name: String,
    override val format: AudioFormat
) : AudioDevice {

    override val frameSize = format.sampleRate.toInt() / 1000 * 20

    private val filters: ListMultimap<AudioFilter.Priority, AudioFilter> = Multimaps.synchronizedListMultimap(
        Multimaps.newListMultimap(HashMap<AudioFilter.Priority, Collection<AudioFilter>>()) { ArrayList() }
    )

    @Throws(DeviceException::class)
    override fun reload() {
        if (!isOpen()) throw DeviceException("Device is not open")

        close()
        open()
    }

    override fun addFilter(filter: AudioFilter, priority: AudioFilter.Priority) {
        // check if filter already exists
        for (filtersPriority in AudioFilter.Priority.values()) {
            val filters: Collection<AudioFilter> = filters[filtersPriority]
            require(!filters.contains(filter)) { "Filter is already exist with priority: $filtersPriority" }
        }
        filters.put(priority, filter)
    }

    override fun removeFilter(filter: AudioFilter) {
        for (filtersPriority in AudioFilter.Priority.values()) {
            val filters: MutableCollection<AudioFilter> = filters[filtersPriority]
            filters.remove(filter)
        }
    }

    override fun getFilters(): Collection<AudioFilter> =
        filters.values()

    override fun processFilters(samples: ShortArray, excludeFilter: Predicate<AudioFilter>?): ShortArray {
        val context = AudioFilterContext(this)

        var samples = samples
        for (filter in filters.values()) {
            if (!filter.isEnabled) continue
            if (filter.supportedChannels > 0 && filter.supportedChannels != context.channels) continue
            if (excludeFilter?.test(filter) == true) continue

            samples = filter.process(context, samples)
        }
        return samples
    }

    @Throws(DeviceException::class)
    protected abstract fun open()
}
